<?xml version="1.0"?>

<launch>
    <arg name="use_rviz" default="true" />

    <!-- for amcl -->    
    <arg name="init_x" default="0.0" />
    <arg name="init_y" default="0.0" />
    <arg name="init_a" default="0.0" />

    <node pkg="tf" type="static_transform_publisher" name="map_to_odom" args="0.0 0.0 0.0 0 0 0.0 /map /odom 1000"/>
    <!-- Map server -->
    
    <node name="map_server" pkg="map_server" type="map_server" args="$(find racecar)/map/mymap.yaml"/>

   

    <!-- ODOMETRY -->
    <!--rf2o_Laser_Odometry-->
    <include file="$(find racecar)/launch/includes/rf2o.launch.xml" />
    <!-- wheel odometry -->
    <include file="$(find encoder_driver)/launch/wheel_odom.launch"/>
    <!-- Robot_Localization -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find racecar)/param/ekf_params.yaml" />
    </node>

    <!-- Localization -->
    <!-- AMCL -->
    <include file="$(find racecar)/launch/includes/amcl.launch.xml">
        <arg name="init_x" value="$(arg init_x)"/>
        <arg name="init_y" value="$(arg init_y)"/>
        <arg name="init_a" value="$(arg init_a)"/>
    </include>

    <!-- Navstack -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base">
        <!-- local planner -->
      
        <param name="base_global_planner" value="navfn/NavfnROS"/>  
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>  
        <rosparam file="$(find racecar)/param/dwa_local_planner_params.yaml" command="load"/>
        
        <!-- costmap layers -->    
        <rosparam file="$(find racecar)/param/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find racecar)/param/global_costmap_params.yaml" command="load"/> 
        <!-- move_base params -->
        <rosparam file="$(find racecar)/param/base_global_planner_params.yaml" command="load"/>
        <rosparam file="$(find racecar)/param/move_base_params.yaml" command="load"/>
    </node>

    <node pkg="racecar" type="car_controller_new" respawn="false" name="car_controller" output="screen">       
       
        <param name="Vcmd" value="1.5" /> <!--speed of car m/s         -->
        <!-- ESC -->
        <param name="baseSpeed" value="150"/> 
        <param name="baseAngle" value="0.0"/> 
        <param name="Angle_gain_p" value="-3.0"/> 
        <param name="Angle_gain_d" value="-3.0"/> 
        <param name="Lfw" value="1.5"/> 
        <param name="vp_max_base" value="200"/> 
        <param name="vp_min" value="200"/> 

    </node>

    <!-- Rviz -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find racecar)/rviz/amcl.rviz" if="$(arg use_rviz)" />
</launch>
